src/Microsoft.DotNet.Tools.Repl.Csi/Program.cs

```diff
@@ -19,21 +24,128 @@ public static int Main(string[] args)
+                    var runtimeAssemblyPath = runtimeAssembly.ResolvedPath;
+
+                    if (!runtimeDependencies.Contains(runtimeAssemblyPath))
+                    {
+                        runtimeDependencies.Add(runtimeAssemblyPath);
+                    }
+                }
+            }
+
+            return runtimeDependencies;
+        }
+
+        private static string CreateResponseFile(ProjectContext projectContext, string buildConfiguration, string tempOutputDir)
+        {
+            var outputFileName = projectContext.ProjectFile.Name;
+            var outputFilePath = Path.Combine(tempOutputDir, $"{outputFileName}{Constants.DynamicLibSuffix}");
+            var projectResponseFilePath = Path.Combine(tempOutputDir, $"dotnet-repl.{outputFileName}{Constants.ResponseFileSuffix}");
+
+            var runtimeDependencies = GetRuntimeDependencies(projectContext, buildConfiguration);
+
+            using (var fileStream = new FileStream(projectResponseFilePath, FileMode.Create))
+            {
+                var streamWriter = new StreamWriter(fileStream);
+
+                streamWriter.WriteLine($"/r:\"{outputFilePath}\"");
+
+                foreach (var projectDependency in runtimeDependencies)
+                {
+                    streamWriter.WriteLine($"/r:\"{projectDependency}\"");
+                }
+
+                streamWriter.Flush();
+            }
+
+            return projectResponseFilePath;
+        }
+
+        private static int Run(string script, IEnumerable<string> targetFrameworks, string buildConfiguration, bool preserveTemporaryOutput, string projectPath)
         {
             var corerun = Path.Combine(AppContext.BaseDirectory, Constants.HostExecutableName);
             var csiExe = Path.Combine(AppContext.BaseDirectory, "csi.exe");
-            var csiArgs = string.IsNullOrEmpty(scriptOpt) ? "-i" : scriptOpt;
-            var result = Command.Create(csiExe, csiArgs)
+            var csiArgs = new StringBuilder();
+
+            var tempOutputDir = string.Empty;
+
+            if (!string.IsNullOrWhiteSpace(projectPath))
+            {
+                var projectContext = GetProjectContext(targetFrameworks, projectPath);
+
+                var compileResult = CompileProject(projectContext, buildConfiguration, out tempOutputDir);
+
+                if (compileResult.ExitCode != 0)
+                {
+                    return compileResult.ExitCode;
+                }
+
+                string responseFile = CreateResponseFile(projectContext, buildConfiguration, tempOutputDir);
+                csiArgs.Append($"@\"{responseFile}\" ");
+            }
+
+            csiArgs.Append(string.IsNullOrEmpty(script) ? "-i" : script);
+
+            var result = Command.Create(csiExe, csiArgs.ToString())
                 .ForwardStdOut()
                 .ForwardStdErr()
                 .Execute();
+
+            if (!string.IsNullOrWhiteSpace(tempOutputDir) && !preserveTemporaryOutput)
```

Comment:
`tempOutputDir` should never be spaces only. And perhaps initialize to `null` so this check can be simplified to `tempOutputDir != null`.